### PR TITLE
Ctcp 3601 - Functional error pagination

### PR DESCRIPTION
**Pull request by ajchilts**
Paging has been implemented in this PR by slicing the functional error list returned from the API after each subsequent call. 

As this slice takes place in memory and keeps the view state light I do not believe we will encounter any performance hits. 

I do however feel that we should performance test this behaviour. 

 

Note: Each time the user pages an API call will be made and we will slice the results and render only a subset of data. If we want/need to make this more optimum we can add a mongo cache (session cache) to the front end with a 15 minute expiry. The behaviour would then be as follows...

 

- On page load attempt to fetch any functional errors from the cache
- If in cache then slice (page) and return to the view
- If not in cache then fetch from API, cache and then slice and return to the view
- On paging repeat these steps. 

This would prevent API calls when paging however, would require more dev effort. 

I believe it is unlikely that a user will have a vast number of errors. I also believe that the API could handle the load of users paging. This might be something we want to consider however. 

**Screenshots**

Screenshot 2023-08-01 at 16 11 17
Screenshot 2023-08-01 at 16 11 28
Screenshot 2023-08-01 at 16 11 35
Screenshot 2023-08-01 at 16 11 42

### CODE REVIEW

File: app/viewModels/P5/departure/ReviewDepartureErrorsP5ViewModel.scala [hunk 1]
```diff
@@ -63,12 +63,20 @@ object ReviewDepartureErrorsP5ViewModel {
 
     def apply(
       ie056MessageData: IE056MessageData,
-      lrn: String
+      lrn: String,
+      start: Int,
```
Comment (jackheslop96) [inline app/viewModels/P5/departure/ReviewDepartureErrorsP5ViewModel.scala:67]: Maybe call this `currentPage`? (assuming that's what it is?)

File: app/viewModels/P5/departure/ReviewDepartureErrorsP5ViewModel.scala [hunk 2]
```diff
@@ -63,12 +63,20 @@ object ReviewDepartureErrorsP5ViewModel {
...
+      numberOfErrors: Int
```
Comment (jackheslop96) [inline app/viewModels/P5/departure/ReviewDepartureErrorsP5ViewModel.scala:68]: Maybe call this `numberOfErrorsPerPage`? (assuming that's what it is?)

File: app/viewModels/P5/departure/ReviewDepartureErrorsP5ViewModel.scala [hunk 3]
```diff
@@ -63,12 +63,20 @@ object ReviewDepartureErrorsP5ViewModel {
...
     )(implicit messages: Messages, ec: ExecutionContext, hc: HeaderCarrier): Future[ReviewDepartureErrorsP5ViewModel] = {
-      val helper = new RejectionMessageP5MessageHelper(ie056MessageData.functionalErrors, referenceDataService)
+      val paged = ie056MessageData.functionalErrors
+        .sortBy(
+          x => x.errorCode
+        )
+        .slice((start - 1) * numberOfErrors, ((start - 1) * numberOfErrors) + numberOfErrors)
```
Comment (jackheslop96) [inline app/viewModels/P5/departure/ReviewDepartureErrorsP5ViewModel.scala:74]: Then could have
```
val paged = {
  val start = (currentPage - 1) * numberOfErrorsPerPage
  ie056MessageData.functionalErrors
    .sortBy(_.errorCode)
    .slice(start, start + numberOfErrorsPerPage)
}
```
Maybe makes the logic a bit clearer?

**Review by jackheslop96 — DISMISSED** [review]

**Review by jackheslop96 — APPROVED** [review]

File: conf/application.conf [hunk 4]
```diff
@@ -145,7 +145,7 @@ microservice {
     features {
       isGuaranteeBalanceEnabled = true
       phase5Enabled {
-        departure = false
+        departure = true
```
Comment (kenneth-rayner) [inline conf/application.conf:148]: revert to false?

**Review by kenneth-rayner — APPROVED** [review]

**Review by vajarova — APPROVED** [review]

**Pull request merged**